### ISSUE TITLE
BLOCKER: Docker forgets about volumes created with plugin drivers after service restart

### ISSUE THREAD

**Issue by nickwilliams-eventbrite** [docker/docker#19987]
I created a plugin volume driver that stores volumes on a remote file system. I then created a standalone volume with the following command:

    $ docker volume create --driver=my-driver --name=geo-vol --opt BindDirectory=/work/docker-dev/volumes/geo-vol
    geo-vol

And it works:

    $ docker volume inspect geo-vol
    [
        {
            "Name": "geo-vol",
            "Driver": "my-driver",
            "Mountpoint": "/work/docker-dev/volumes/geo-vol"
        }
    ]
    $ ls -al /work/docker-dev/volumes/geo-vol/
    total 0
    drwxrwxrwx 1 vagrant vagrant  68 Feb  3 19:19 .
    drwxrwxrwx 1 vagrant vagrant 680 Feb  3 19:19 ..

I am able to mount the volume in multiple containers and read and write data from all the containers. It's a great way to share data between containers! But there's a problem...

    $ docker volume inspect geo-vol
    [
        {
            "Name": "geo-vol",
            "Driver": "my-driver",
            "Mountpoint": "/work/docker-dev/volumes/geo-vol"
        }
    ]
    $ sudo service docker restart
    docker stop/waiting
    docker start/running, process 27923
    $ docker volume inspect geo-vol
    Error response from daemon: no such volume
    $ docker volume ls | grep geo
    [no output]

When restarting, Docker completely forgot that the `geo-vol` volume exists. But my driver didn't. When trying to re-create it:

    $ docker volume create --driver=my-driver --name=geo-vol --opt BindDirectory=/work/docker-dev/volumes/geo-vol
    Error response from daemon: The volume geo-vol already exists.

Looking through the HTTP logs for my driver, it's definitely working correctly. When it works (before Docker restarts), `docker volume inspect` calls my driver (as it should) to get the volume path:

    Pid-28176 Thread-2 localhost-socket - - [04/Feb/2016 03:00:18] "POST /VolumeDriver.Path HTTP/1.1" 200 62 '{"Name":"geo-vol"}'

But when it doesn't work (after Docker restarts), `docker volume inspect` doesn't bother to call my driver. So the problem is definitely that Docker has forgotten the volume during restart, not that my driver is doing something wrong.

This is a major blocker for us. I'm going to have to revert our Docker environment to using data volume containers until this is fixed. :-(

**Comment by nickwilliams-eventbrite** [comment]
I'm running on VirtualBox. Here are my Docker and Ubuntu versions information: 

    $ docker version
    Client:
     Version:      1.9.0
     API version:  1.21
     Go version:   go1.4.3
     Git commit:   76d6bc9
     Built:        Tue Nov  3 19:20:09 UTC 2015
     OS/Arch:      linux/amd64

    Server:
     Version:      1.9.0
     API version:  1.21
     Go version:   go1.4.3
     Git commit:   76d6bc9
     Built:        Tue Nov  3 19:20:09 UTC 2015
     OS/Arch:      linux/amd64

    $ lsb_release -a
    No LSB modules are available.
    Distributor ID:	Ubuntu
    Description:	Ubuntu 14.04 LTS
    Release:	14.04
    Codename:	trusty

**Comment by cpuguy83** [comment]
This is fixed in docker 1.10
Please make sure to update your volume driver to support the new RPC calls.

**Comment by nickwilliams-eventbrite** [comment]
Is there documentation for the new RPC calls? https://docs.docker.com/engine/extend/plugins_volume/ mentions nothing about them. Are the new RCP calls a breaking change? Will my volume driver stop working under 1.10 without my making changes?

**Comment by cpuguy83** [comment]
latest docs are here: https://github.com/docker/docker/blob/master/docs/extend/plugins_volume.md
Since 1.10 isn't out yet (should be happening tomorrow) these are not on docs.docker.com yet.

If you are using a volume plugin that does not support the new calls, it is likely to cause unexpected behavior.

**Comment by nickwilliams-eventbrite** [comment]
Thanks. I see what changed (mostly, what was added). I look forward to trying out 1.10. Out of curiosity, one of the changes confused me, so I thought I'd ask. I noticed that the non-error response for all the calls used to be `"Err": null`, but now it's supposed to be `"Err": ""`. Is it really a necessity for me to change the driver in this way? Will it work either way? Will 1.9 work with blank strings instead of `null`, or will 1.10 work with `null` instead of blank strings? If it's really a breaking change, it seems like a pretty meaningless reason to introduce a breaking change. :-/
